WD-Firmware/comrv/comrv_entry.S

```diff
@@ -0,0 +1,188 @@
+   t3 - points to current stack top (per task if OS involved)
+   t4 - points to the free pool of stack frames
+   t5 - holds the token we need to invoke (can be an actual address if callee
+        is a non-overlay function) */
+comrv_entry:
+    /* allocate a new stack frame from the stack frame pool - we unlink the next
+       free frame from the stack frame pool (t4) and assign its address to t2.
+       t4 is updated with the next free stack frame */
+    /* disable interrupts */
+    m_COMRV_DISABLE_INTS
+    /* save the address of the next free stack frame in t2 - new stack frame
+       for this engine call */
+    mv      t2, t4
+    /* load to t4 the offset to the previous free stack frame */
+    lh      t4, D_COMRV_PREV_STACK_FRAME(t2)
+    /* set in t4 the address of the next free stack frame - will be used in
+        the next engine call */
+    add     t4, t2, t4
+    /* enable interrupts */
+    m_COMRV_ENABLE_INTS
+    /* at this point t4 points to the next free frame, t3 points to the top
+       of the com-rv stack and t2 holds the allocated stack frame. */
+    /* link the allocated frame to its stack - we link the allocated frame
+       to the stack pointed by t3. */
+    /* calc the offset from the new stack frame (t2) to previous stack frame (t3) */
+    sub     t3, t3, t2
+    /* in current stack frame (t2) save the offset to previous stack frame */
+    sh      t3, D_COMRV_PREV_STACK_FRAME(t2)
+    /* link allocated stack frame */
+    mv      t3, t2
+    /* save callee token */
+    m_STORE t5, D_COMRV_CALLEE_TOKEN_STACK_OFFSET(t3)
+    /* save the return address */
+    m_STORE ra, D_COMRV_CALLER_RA_STACK_OFFSET(t3)
+    /* check if callee token is an overlay token (in t5 bit0 is set) */
+    andi    t2, t5, D_COMRV_TOKEN_INDICATION
+    /* if t2 is 0 it means callee is a non overlay function so we can
+       skip search and load */
+    beqz    t2, comrv_invoke_callee
+    /* change sp for temporary storage */
+    m_ADDI  sp, sp, -D_COMRV_STACK_FRAME_SIZE
+    /* temporary save a0 in the next free stack frame */
+    m_STORE a0, D_COMRV_A0_REG_STACK_OFFSET_TMP(sp)
+    /* current comrv stack frame is needed when calling a new overlay function so */
+    mv      a0, t3
+    /* Add the indication we are calling a new overlay function
+       this indication is required to calculate the correct offset
+       when returning to an overlay function */
+    ori     a0, a0, D_COMRV_PROFILING_INVOKE_VAL
+    /* search if the token (t5) is already loaded */
+    jal     comrvGetAddressFromToken
+    /* save the loaded function address in t5 (instead of the token) */
+    mv      t5, a0
+    /* restore a0 - may contain an actual callee function argument */
+    m_LOAD  a0, D_COMRV_A0_REG_STACK_OFFSET_TMP(sp)
+    /* adjust back sp - memory not needed anymore */
+    m_ADDI  sp, sp, D_COMRV_STACK_FRAME_SIZE
+comrv_invoke_callee:
+    /* call the function */
+    jalr    t5
+comrv_skip_thunk:
```

Comment:
change to igonr_callee_thunk_stack_frame

Reply:
I used comrv_igonr_caller_thunk_stack_frame as when returning, the thunk is the caller and not the callee